fix huggingface tokenizer default length function

keshavshrikant opened the pull request:
#30184

keshavshrikant committed:
fix huggingface tokenizer default length function

ccurme reviewed (commented):
Hesitant to do this for two reasons:

1. `.encode` is [documented](https://github.com/huggingface/transformers/blob/94ae1ba5b55e79ba766582de8a199d8ccf24a021/src/transformers/tokenization_utils_base.py#L2642) as "Same as doing `self.convert_tokens_to_ids(self.tokenize(text))`". Is this a bug in that tokenizer?
2. Would this change behavior out from under users of bge-m3? All HF tokenizers?

Is this just an issue with bge-m3, or with all HF tokenizers?

keshavshrikant commented:
`.encode` has a parameter `add_special_tokens` which defaults to `True`. Changing it to false makes the behaviour same as `.tokenize`

Tried it out for `sentence-transformers/all-mpnet-base-v2` and got the same output

```python
from transformers import AutoTokenizer
tokenizer_mpnet = AutoTokenizer.from_pretrained("sentence-transformers/all-mpnet-base-v2")
tokenizer_mpnet.tokenize("What is my name")
['what', 'is', 'my', 'name']

tokenizer_mpnet.encode("What is my name")
[0, 2058, 2007, 2030, 2175, 2]

tokenizer_mpnet.encode("What is my name", add_special_tokens=False)
[2058, 2007, 2030, 2175]
```
This behaviour is consistent with the bug I had raised #30184 in that changing the `add_special_tokens` from `True` to `False` makes the chunks exactly same.

ccurme reviewed (commented):
Thanks for the reply. I have a few questions about this still.

Can you clarify, in the issue when you say
> I observed that an overwhelming majority of the chunks were nowhere near the 256 tokens length

How are you measuring the length of the chunks (in tokens)?

I ran a rough experiment using a few tokenizers, using `.split()` as a proxy for tokenizing (since `tokenizer.encode` is itself in question here).
```python
from langchain_text_splitters import CharacterTextSplitter
from transformers import AutoTokenizer, GPT2TokenizerFast


# https://github.com/langchain-ai/langchain/blob/master/docs/docs/how_to/state_of_the_union.txt
with open("state_of_the_union.txt") as f:
    state_of_the_union = f.read()

tokenizer_params = {"chunk_size": 100, "chunk_overlap": 0}

splitters = {
    "tiktoken": CharacterTextSplitter.from_tiktoken_encoder(
        encoding_name="cl100k_base", **tokenizer_params,
    ),
    "GPT2TokenizerFast": CharacterTextSplitter.from_huggingface_tokenizer(
        GPT2TokenizerFast.from_pretrained("gpt2"),
        **tokenizer_params,
    ),
    "GPT2, Autotokenizer": CharacterTextSplitter.from_huggingface_tokenizer(
        AutoTokenizer.from_pretrained("gpt2"),
        **tokenizer_params,
    ),
    "all-mpnet-base-v2": CharacterTextSplitter.from_huggingface_tokenizer(
        AutoTokenizer.from_pretrained(
            "sentence-transformers/all-mpnet-base-v2"
        ),
        **tokenizer_params,
    ),
    "bert": CharacterTextSplitter.from_huggingface_tokenizer(
        AutoTokenizer.from_pretrained("google-bert/bert-base-cased"),
        **tokenizer_params,
    ),
    "BGE-M3": CharacterTextSplitter.from_huggingface_tokenizer(
        AutoTokenizer.from_pretrained("BAAI/BGE-M3"),
        **tokenizer_params,
    ),
}
for name, splitter in splitters.items():
    texts = splitter.split_text(state_of_the_union)
    print(f"{name:<20} {len(texts[0].split()):>5}")
```
Results on the `master` branch suggest the issue is with `BGE-M3`:
```
tiktoken                64
GPT2TokenizerFast       64
GPT2, Autotokenizer     64
all-mpnet-base-v2       64
bert                    64
BGE-M3                  51
```
Here are the results on your branch:
```
tiktoken                64
GPT2TokenizerFast       64
GPT2, Autotokenizer     64
all-mpnet-base-v2       76
bert                    64
BGE-M3                  64
```
So, looks pretty good, but wondering if we're trading for an issue with `all-mpnet-base-v2`. Does the change there make sense to you?

keshavshrikant commented:
I am using the `tokenize` function to compute length, but `split()` works as a proxy. I made the following changes for a comparison:

1. Replaced `\n` with a `blankspace`
2. Used `RecursiveCharacterTextSplitter` instead of `CharacterTextSplitter` since that's the one I am facing an issue with

```python
import re
from langchain_text_splitters import CharacterTextSplitter, RecursiveCharacterTextSplitter
from transformers import AutoTokenizer, GPT2TokenizerFast


# https://github.com/langchain-ai/langchain/blob/master/docs/docs/how_to/state_of_the_union.txt
with open("state_of_the_union.txt") as f:
    state_of_the_union = f.read()

state_of_the_union = re.sub(r'\n', ' ', state_of_the_union)
tokenizer_params = {"chunk_size": 100, "chunk_overlap": 0}

splitters = {
    "tiktoken": RecursiveCharacterTextSplitter.from_tiktoken_encoder(
        encoding_name="cl100k_base", **tokenizer_params,
    ),
    "GPT2TokenizerFast": RecursiveCharacterTextSplitter.from_huggingface_tokenizer(
        GPT2TokenizerFast.from_pretrained("gpt2"),
        **tokenizer_params,
    ),
    "GPT2, Autotokenizer": RecursiveCharacterTextSplitter.from_huggingface_tokenizer(
        AutoTokenizer.from_pretrained("gpt2"),
        **tokenizer_params,
    ),
    "all-mpnet-base-v2": RecursiveCharacterTextSplitter.from_huggingface_tokenizer(
        AutoTokenizer.from_pretrained(
            "sentence-transformers/all-mpnet-base-v2"
        ),
        **tokenizer_params,
    ),
    "bert": RecursiveCharacterTextSplitter.from_huggingface_tokenizer(
        AutoTokenizer.from_pretrained("google-bert/bert-base-cased"),
        **tokenizer_params,
    ),
    "BGE-M3": RecursiveCharacterTextSplitter.from_huggingface_tokenizer(
        AutoTokenizer.from_pretrained("BAAI/BGE-M3"),
        **tokenizer_params,
    ),
}
for name, splitter in splitters.items():
    texts = splitter.split_text(state_of_the_union)
    print(f"{name:<20} {len(texts[0].split()):>5}")
```

Here are the default results:
```
tiktoken                70
GPT2TokenizerFast       70
GPT2, Autotokenizer     70
all-mpnet-base-v2       19
bert                    19
BGE-M3                  18
```

Here are the results from my branch
```
tiktoken                70
GPT2TokenizerFast       70
GPT2, Autotokenizer     70
all-mpnet-base-v2       79
bert                    75
BGE-M3                  64
```

With `\n` in place, the results are exactly similar to what you posted.

Additionally, alluding to the issue I had raised in #30184, there are no issues with `tiktoken encoder`. Also,  the `gpt2 tokenizer` from `AutoTokenizer.from_pretrained` or `GPT2TokenizerFast.from_pretrained` doesn't seem to be counting special characters (like `start` and `end` for example). So `encode` and `tokenize` end up giving the same length. 

```python
gpt2_tokenizer = AutoTokenizer.from_pretrained("gpt2") # GPT2TokenizerFast.from_pretrained("gpt2")
s = "What is my name"
print(gpt2_tokenizer.encode(s))
print(gpt2_tokenizer.tokenize(s))
```
Outputs:
```
Encode Output: [2061, 318, 616, 1438]
Tokenize Output: ['What', 'Ġis', 'Ġmy', 'Ġname']
```

I think using `tokenize` as the length function will make the code work consistently across tokenizers. With `encode` as the length function, we run the risk of counting those special characters for each split before merging, but I don't think that's the case with `tokenize`.
